File: codes/naeun/876_middleofthelinkedlist_1.java
class Solution {
    public ListNode middleNode(ListNode head) {
        ListNode s1=head;
        ListNode s2=head;

        while(s2!=null && s2.next!=null){
            s1=s1.next;
            s2=s2.next.next;
        }
        return s1;
    }
}

File: codes/beomjin/876_middleofthelinkedlist_1.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
import scala.annotation.tailrec

object Solution {
  def middleNode(head: ListNode): ListNode = {
    val slow, fast = head

    @tailrec
    def findMid(s: ListNode, f: ListNode): ListNode =
      (s, f) match
        case (null, null) => null
        case (s, null) => s
        case (s, n) =>
          if (f.next == null) then findMid(s, null)
          else findMid(s.next, f.next.next)

      findMid(head, head)
  }
}